src/middlewares/csrf.ts

```diff
@@ -0,0 +1,94 @@
+import type { FreshContext } from "../context.ts";
+
+export type IsAllowedOriginHandler = (
+  origin: string,
+  context: FreshContext,
+) => boolean;
+export interface CSRFOptions {
+  origin?: string | string[] | IsAllowedOriginHandler;
+}
+
+const isSafeMethodRe = /^(GET|HEAD)$/;
+const isRequestedByFormElementRe =
+  /^\b(application\/x-www-form-urlencoded|multipart\/form-data|text\/plain)\b/i;
+
+/**
+ * CSRF Protection Middleware for Fresh.
+ * Adapted from Hono's CORS middleware.
+ * Would like to express my gratitude to all the developers of Hono.
+ *
+ * @param {CSRFOptions} [options] - The options for the CSRF protection middleware.
+ * @param {string|string[]|(origin: string, context: Context) => boolean} [options.origin] - Specify origins.
+ * @returns {MiddlewareHandler} The middleware handler function.
```

Comment:
```suggestion
 * @param options The options for the CSRF protection middleware.
 * @returns The middleware handler function.
```
Type information isn't needed in JSDocs for TypeScript.